File: test/instr/EnumTransformTest.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package redex;

import java.util.Random;
import javax.annotation.Nullable;
import org.junit.Test;
import static org.fest.assertions.api.Assertions.assertThat;

// POSTCHECK-LABEL: class: redex.$EnumUtils
// POSTCHECK-NEXT: Access flags: (PUBLIC, FINAL)
// POSTCHECK-NEXT: Superclass: java.lang.Object
// POSTCHECK: (PRIVATE, STATIC, FINAL) $VALUES:java.lang.Integer[]
// POSTCHECK: (PUBLIC, STATIC, FINAL) f0:java.lang.Integer
// POSTCHECK: (PUBLIC, STATIC, FINAL) f1:java.lang.Integer
// POSTCHECK: (PUBLIC, STATIC, FINAL) f2:java.lang.Integer
// POSTCHECK: (PUBLIC, STATIC, FINAL) f3:java.lang.Integer
// POSTCHECK-NOT: (PUBLIC, STATIC, FINAL) f4:java.lang.Integer

// This enum class contains some user defined static methods and it's never
// being casted to other types. It will be transformed to a class with only
// static methods by OptimizeEnumsPass.
// CHECK-LABEL: class: redex.SCORE
// CHECK-NEXT: Access flags:
// PRECHECK-NEXT: Superclass: java.lang.Enum
// POSTCHECK-NEXT: Superclass: java.lang.Object
enum SCORE {
  ONE(11, "UNO"),
  TWO(12, "DOS"),
  THREE(13, null);

  SCORE(int data, String str) {
    myOtherField = str;
    myField = data;
    scoreToWin = 101;
    constantString = "IDoNotChange";
  }

  static final SCORE DEFAULT = ONE;
  static final SCORE[] array = values();
  static int number = 0;

  // Primitive or String instance fields are (usually) safe.
  // POSTCHECK-DAG: method: direct redex.SCORE.redex$OE$get_myField:(java.lang.Integer)int
  int myField;
  // POSTCHECK-DAG: method: direct redex.SCORE.redex$OE$get_myOtherField:(java.lang.Integer)java.lang.String
  String myOtherField;
  int scoreToWin;
  String constantString;

  // POSTCHECK-DAG: method: direct redex.SCORE.increase$REDEX${{.*}}:(java.lang.Integer)java.lang.Integer
  public static @Nullable SCORE increase(@Nullable SCORE score) {
    if (score == null) {
      return null;
    }
    switch (score) {
    case ONE:
      return TWO;
    case TWO:
    case THREE:
      return THREE;
    }
    return null;
  }

  private SCORE instance_direct() { return this; }

  // Virtual methods are safe.
  // POSTCHECK-DAG: method: direct redex.SCORE.is_max$REDEX${{.*}}:(java.lang.Integer)boolean
  public boolean is_max() { return instance_direct() == THREE; }

  // POSTCHECK-DAG: method: direct redex.SCORE.toString$REDEX${{.*}}:(java.lang.Integer)java.lang.String
  public String toString() {
    return this.myOtherField;
  }

  // POSTCHECK-DAG: method: direct redex.SCORE.getString$REDEX${{.*}}:(java.lang.Integer)java.lang.String
  public String getString() {
    return this.myOtherField;
  }

  // POSTCHECK-DAG: method: direct redex.SCORE.cast_to_enum_array$REDEX${{.*}}:(java.lang.Integer)java.lang.Integer[]
  public SCORE[] cast_to_enum_array() {
    Object result = SCORE.values();
    return (SCORE[]) result;
  }
}

interface Intf {
  int make(SCORE score);
  int make(int i);
}

// Some usages of SCORE.
class A {
  // As parameter.
  // POSTCHECK: method: virtual redex.A.ha$REDEX${{.*}}:(java.lang.Integer)int
  public int ha(SCORE score) { return 1; }
}

abstract class B extends A implements Intf {
  // As return value.
  // POSTCHECK: method: virtual redex.B.haha$REDEX${{.*}}:()java.lang.Integer
  public SCORE haha() { return SCORE.THREE; }
  public int make(int i) { return i * 7; }
  int ha(int i) { return 11; }
}

// CHECK-LABEL: class: redex.C
class C extends B {
  static @Nullable SCORE s_score = SCORE.ONE;
  // PRECHECK: (STATIC) array:redex.SCORE[][]
  // POSTCHECK: (STATIC) array$REDEX${{.*}}:java.lang.Integer[][]
  static SCORE[][] array;
  @Nullable SCORE i_score = SCORE.ONE;
  @Nullable Object i_obj = null;

  public void set(SCORE score) { i_score = score; }
  public @Nullable SCORE get() { return i_score; }

  public int make(SCORE score) { return score.ordinal() * 5; }
}

// This enum class only contains several Enum objects.
// It will be optimized by OptimizeEnumsPass.
// PRECHECK: class: redex.PURE_SCORE
// PRECHECK-NEXT: Access flags:
// PRECHECK-NEXT: Superclass: java.lang.Enum
// POSTCHECK-NOT: class redex.PURE_SCORE
enum PURE_SCORE {
  ONE,
  TWO,
  THREE,
  FOUR,
  FIVE;
}

/* Some enums that are unsafe to be transformed. */
// CHECK: class: redex.USED_IN_UNSAFE_CONSTRUCTOR
// CHECK-NEXT: Access flags:
// CHECK-NEXT: Superclass: java.lang.Enum
enum USED_IN_UNSAFE_CONSTRUCTOR {
  ONE;
}
class HasUnsafeConstructor {
  HasUnsafeConstructor(USED_IN_UNSAFE_CONSTRUCTOR[] e) {}
  HasUnsafeConstructor(Integer[] e) {}
}
// CHECK: class: redex.MODIFIES_INSTANCE_FIELD
// CHECK-NEXT: Access flags:
// CHECK-NEXT: Superclass: java.lang.Enum
enum MODIFIES_INSTANCE_FIELD {
  ONE(1234);
  int myField;
  MODIFIES_INSTANCE_FIELD(int data) {
    myField = data;
  }
  public void modify() {
    myField++;
  }
}
// CHECK: class: redex.CAST_WHEN_RETURN
// CHECK-NEXT: Access flags:
// CHECK-NEXT: Superclass: java.lang.Enum
enum CAST_WHEN_RETURN {
  ONE;
  public static Enum[] method() {
    CAST_WHEN_RETURN[] array = new CAST_WHEN_RETURN[1];
    array[0] = ONE;
    return array;
  }
}
// CHECK: class: redex.CAST_THIS_POINTER
// CHECK-NEXT: Access flags:
// CHECK-NEXT: Superclass: java.lang.Enum
enum CAST_THIS_POINTER {
  ONE;
  public static void cast_this_method() {
    EnumHelper.inlined_method(CAST_THIS_POINTER.ONE);
  }
}
// CHECK: class: redex.CAST_THIS_POINTER_2
// CHECK-NEXT: Access flags:
// CHECK-NEXT: Superclass: java.lang.Enum
enum CAST_THIS_POINTER_2 {
  ONE;
  public String cast_this_method() {
    return super.toString();
  }
}
// CHECK: class: redex.CAST_PARAMETER
// CHECK-NEXT: Access flags:
// CHECK-NEXT: Superclass: java.lang.Enum
enum CAST_PARAMETER {
  ONE;
  public static <E extends Enum<E>> void method_accepts_enum_arg(
      @Nullable Enum<E> o) {}
  public static void method() { method_accepts_enum_arg(ONE); }
}
// CHECK: class: redex.USED_AS_CLASS_OBJECT
// CHECK-NEXT: Access flags:
// CHECK-NEXT: Superclass: java.lang.Enum
enum USED_AS_CLASS_OBJECT {
  ONE;
  public static <T> void method(Class<T> c) {}
  public static void method() { method(USED_AS_CLASS_OBJECT.class); }
}
// CHECK: class: redex.USED_IN_INSTANCE_OF
// CHECK-NEXT: Access flags:
// CHECK-NEXT: Superclass: java.lang.Enum
enum USED_IN_INSTANCE_OF {
  ONE;
  public static void method(boolean b) {}
  public static void method() { method(ONE instanceof USED_IN_INSTANCE_OF); }
}
// CHECK: class: redex.CAST_CHECK_CAST
// CHECK-NEXT: Access flags:
// CHECK-NEXT: Superclass: java.lang.Enum
enum CAST_CHECK_CAST { ONE; }
// CHECK: class: redex.CAST_ISPUT_OBJECT
// CHECK-NEXT: Access flags:
// CHECK-NEXT: Superclass: java.lang.Enum
enum CAST_ISPUT_OBJECT { ONE; }
// CHECK: class: redex.CAST_APUT_OBJECT
// CHECK-NEXT: Access flags:
// CHECK-NEXT: Superclass: java.lang.Enum
enum CAST_APUT_OBJECT { ONE; }
// CHECK: class: redex.ENUM_TYPE_1
// CHECK-NEXT: Access flags:
// CHECK-NEXT: Superclass: java.lang.Enum
enum ENUM_TYPE_1 { ONE; }
// CHECK: class: redex.ENUM_TYPE_2
// CHECK-NEXT: Access flags:
// CHECK-NEXT: Superclass: java.lang.Enum
enum ENUM_TYPE_2 {
  ONE,
  TWO;
  public static void test_join() {
    Enum obj = null;
    Random random = new Random();
    int pos_rand = random.nextInt() & Integer.MAX_VALUE;
    int selector = pos_rand % 2;
    if (selector == 0) {
      obj = ENUM_TYPE_1.ONE;
    } else if (selector == 1 || selector == -1) {
      obj = ENUM_TYPE_2.TWO;
    } else if (selector > 1) {
      obj = Enum.valueOf(ENUM_TYPE_1.class, "NotReach");
    }
    // obj may be Enum, ENUM_TYPE_1 or ENUM_TYPE_2
    int res = obj.ordinal();
    if (selector == 0) {
      assertThat(res).isEqualTo(0);
    } else if (selector == 1) {
      assertThat(res).isEqualTo(1);
    }
  }
}

// CHECK: class: redex.CAST_ENUM_ARRAY_TO_OBJECT
// CHECK-NEXT: Access flags:
// CHECK-NEXT: Superclass: java.lang.Enum
enum CAST_ENUM_ARRAY_TO_OBJECT {
  ONE;
  public static String method() {
    StringBuilder sb = new StringBuilder();
    sb.append(CAST_ENUM_ARRAY_TO_OBJECT.values());
    sb.append(String.valueOf(CAST_ENUM_ARRAY_TO_OBJECT.values()));
    return sb.toString();
  }
}

class EnumHelper {
  static void inlined_method(Enum e) { int a = e.hashCode(); }

  class Cache<T> {
    @Nullable T e = (T) null;
    public @Nullable T get() { return e; }
  }
  Cache<CAST_CHECK_CAST> mCache = new Cache();
  public CAST_CHECK_CAST check_cast_method() { return mCache.get(); }

  Object i_obj;
  static Enum s_obj;
  public void put_method() {
    i_obj = CAST_ISPUT_OBJECT.ONE;
    s_obj = CAST_ISPUT_OBJECT.ONE;
  }

  public void aput_method() {
    Object[] array = new Object[10];
    array[0] = CAST_APUT_OBJECT.ONE;
    s_obj = CAST_APUT_OBJECT.ONE;
  }

  public static int my_ordinal(SCORE score) { return score.ordinal(); }

  public static <T> T notEscape(T obj) { return obj; }
}

/* Test cases */
public class EnumTransformTest {
  @Test
  public void test_interface() {
    C c = new C();
    // Method with the same name but different type of argument.
    assertThat(c.make(SCORE.TWO)).isEqualTo(5);
    assertThat(c.make(1)).isEqualTo(7);
    assertThat(c.ha(0)).isEqualTo(11);
    assertThat(c.ha(null)).isEqualTo(1);
    assertThat(c.haha().ordinal()).isEqualTo(SCORE.THREE.ordinal());
  }

  // SCORE.ordinal() is transformed.
  @Test
  public void test_increase() {
    SCORE score = SCORE.ONE;
    assertThat(score.ordinal()).isEqualTo(SCORE.ONE.ordinal());
    assertThat(score.ordinal()).isEqualTo(0);
    score = SCORE.increase(score);
    assertThat(score.ordinal()).isEqualTo(1);
    score = SCORE.increase(score);
    assertThat(score.ordinal()).isEqualTo(2);
    score = SCORE.increase(score);
    assertThat(score.ordinal()).isEqualTo(2);
    score = SCORE.increase(null);
    if (score != null) {
      assertThat(true).isEqualTo(false);
    }
  }

  // SCORE.equals(SCORE) is transformed.
  @Test
  public void test_equals() {
    SCORE a = SCORE.ONE;
    assertThat(a.equals(SCORE.ONE)).isEqualTo(true);
    SCORE b = SCORE.increase(a);
    assertThat(a.equals(b)).isEqualTo(false);
  }

  // SCORE.compareTo(SCORE) is transformed.
  @Test
  public void test_compareTo() {
    SCORE a = SCORE.ONE;
    SCORE b = SCORE.increase(a);
    assertThat(a.compareTo(b)).isEqualTo(-1);
    assertThat(b.compareTo(a)).isEqualTo(1);
    assertThat(a.compareTo(a)).isEqualTo(0);
  }

  // SCORE.values() is transformed.
  @Test
  public void test_values() {
    SCORE[] values = SCORE.values();
    assertThat(values.length).isEqualTo(3);
    assertThat(SCORE.array.length).isEqualTo(3);
    for (int i = 0; i < 3; i++) {
      assertThat(values[i].ordinal()).isEqualTo(SCORE.array[i].ordinal());
    }
  }

  @Test
  public void test_string_valueof() {
    SCORE obj;
    int rand = new Random().nextInt();
    if (rand >= 0) {
      obj = SCORE.ONE;
    } else {
      obj = null;
    }
    if (rand >= 0) {
      assertThat(String.valueOf(obj)).isEqualTo(SCORE.ONE.toString());
    } else {
      assertThat(String.valueOf(obj)).isEqualTo("null");
    }
  }

  // NullPointerException.
  @Test(expected = NullPointerException.class)
  public void test_npe() {
    int a = EnumHelper.my_ordinal(null);
  }

  // [isa]get_object, [isa]put_object and new_array instructions are
  // transformed.
  @Test
  public void test_get_put() {
    C.s_score = SCORE.ONE;
    assertThat(C.s_score.ordinal()).isEqualTo(0);

    C c = new C();
    c.set(SCORE.TWO);
    assertThat(c.get().ordinal()).isEqualTo(1);

    SCORE[] scores = new SCORE[1];
    scores[0] = SCORE.THREE;
    assertThat(scores[0].ordinal()).isEqualTo(SCORE.THREE.ordinal());
  }

  // PURE_SCORE class is deleted and PURE_SCORE objects are replaced with
  // Integer objects.
  @Test
  public void test_pure_score() {
    PURE_SCORE a = PURE_SCORE.ONE;
    assertThat(a.ordinal()).isEqualTo(PURE_SCORE.ONE.ordinal());
    PURE_SCORE five = PURE_SCORE.FIVE;
    PURE_SCORE five_2 = PURE_SCORE.FIVE;
    assertThat(five.ordinal()).isEqualTo(five_2.ordinal());
    switch (five) {
      case FIVE:
        break;
      default:
        assertThat(true).isEqualTo(false);
    }
  }

  @Test
  public void test_if() {
    C c = new C();
    if (c.i_obj == c.i_score) {
      assertThat(true).isEqualTo(false);
    }
  }

  @Test
  public void test_toString() {
    C c = new C();
    StringBuilder sb = new StringBuilder();
    c.i_score = null;
    sb.append(c.i_score);
    sb.append(SCORE.ONE.toString());
    sb.append(SCORE.TWO);
    sb.append(SCORE.THREE);
    assertThat(sb.toString()).isEqualTo("nullUNODOSnull");
  }

  @Test(expected = NullPointerException.class)
  public void test_toString_throw() {
    C c = new C();
    c.i_score = null;
    c.i_score.toString();
  }

  enum COUNT { ONE, TWO };

  @Test
  public void test_hashCode() {
    SCORE.ONE.hashCode();
    SCORE.TWO.hashCode();
    SCORE.THREE.hashCode();
    // Test an enum that doesn't directly call `Enum.toString()`.
    COUNT.ONE.hashCode();
    COUNT.TWO.hashCode();
  }

  @Test
  public void test_valueOf() {
    String name = SCORE.ONE.name();
    SCORE s = SCORE.valueOf(name);
    assertThat(s.ordinal()).isEqualTo(SCORE.ONE.ordinal());
    s = SCORE.valueOf("TWO");
    assertThat(s.ordinal()).isEqualTo(SCORE.TWO.ordinal());
    s = SCORE.valueOf("THREE");
    assertThat(s.ordinal()).isEqualTo(SCORE.THREE.ordinal());
  }

  @Test(expected = IllegalArgumentException.class)
  public void test_valueOf_exception() {
    SCORE.valueOf("ZERO");
  }

  @Test
  public void test_join_with_multitypes() {
    ENUM_TYPE_2.test_join();
  }

  @Test
  public void test_non_escape_invocation() {
    SCORE one = SCORE.ONE;
    SCORE obj = EnumHelper.notEscape(one);
    assertThat(one.ordinal()).isEqualTo(obj.ordinal());
  }

  @Test
  public void virtual_method() {
    assertThat(SCORE.THREE.is_max()).isEqualTo(true);
  }

  @Test
  public void instance_fields() {
    assertThat(SCORE.ONE.myOtherField).isEqualTo("UNO");
    assertThat(SCORE.TWO.myOtherField).isEqualTo("DOS");
    assertThat(SCORE.THREE.myOtherField).isEqualTo(null);
    assertThat(SCORE.ONE.myField).isEqualTo(11);
    assertThat(SCORE.TWO.myField).isEqualTo(12);
    assertThat(SCORE.THREE.myField).isEqualTo(13);
  }

  @Test
  public void null_enum_value() {
    // PRECHECK:  check-cast {{.*}}, redex.SCORE
    // POSTCHECK: check-cast {{.*}}, java.lang.Integer
    SCORE score_obj = (SCORE) null;
    // CHECK: SCORE.increase
    SCORE.increase(null);
    // CHECK: sput-object {{.*}} redex.C.s_score
    C.s_score = null;
    C c_obj = new C();
    // CHECK: iput-object {{.*}} redex.C.i_score
    c_obj.i_score = null;
    // It's possible that a null value is used as Enum and candidate enum
    // type at the same time.
    SCORE null_ptr = null;
    SCORE.increase(null_ptr);
    CAST_PARAMETER.method_accepts_enum_arg(null_ptr);
  }
}

File: test/instr/EnumTransformTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package redex

import org.fest.assertions.api.Assertions.assertThat
import org.junit.Test

// Simple case is optimized.
// CHECK-LABEL: class: redex.A
// CHECK-NEXT: Access flags:
// PRECHECK-NEXT: Superclass: java.lang.Enum
// POSTCHECK-NEXT: Superclass: java.lang.Object
enum class A {
  A0,
  A1,
  A2
}

interface I {
  fun getVal(): Int
}

// Not optimized if implements an interface.
// CHECK-LABEL: class: redex.B
// CHECK-NEXT: Access flags:
// PRECHECK-NEXT: Superclass: java.lang.Enum
// POSTCHECK-NEXT: Superclass: java.lang.Enum
enum class B : I {
  B0,
  B1,
  B2;
  override fun getVal(): Int {
    return 42
  }
}

// Enum w/ Int field is optimized.
// CHECK-LABEL: class: redex.C
// CHECK-NEXT: Access flags:
// PRECHECK-NEXT: Superclass: java.lang.Enum
// POSTCHECK-NEXT: Superclass: java.lang.Object
enum class C(val idx: Int) {
  C0(1),
  C1(2),
  C2(3)
}

// Enum w/ String field is optimized.
// CHECK-LABEL: class: redex.D
// CHECK-NEXT: Access flags:
// PRECHECK-NEXT: Superclass: java.lang.Enum
// POSTCHECK-NEXT: Superclass: java.lang.Object
enum class D(val bname: String) {
  D0("Tool"),
  D1("APC"),
  D2("Puscifer")
}

data class User(val name: String)

// Enum w/ field of custom type is not optimized.
// CHECK-LABEL: class: redex.E
// CHECK-NEXT: Access flags:
// PRECHECK-NEXT: Superclass: java.lang.Enum
// POSTCHECK-NEXT: Superclass: java.lang.Enum
enum class E(val u: User) {
  E0(User("Hi")),
  E1(User("Yo")),
  E2(User("There"))
}

// Not optimized if enum constant is declared with anonymous class w/ overriding method.
// CHECK-LABEL: class: redex.ProtocolState
// CHECK-NEXT: Access flags:
// PRECHECK-NEXT: Superclass: java.lang.Enum
// POSTCHECK-NEXT: Superclass: java.lang.Enum
enum class ProtocolState {
  WAITING {
    override fun signal(): ProtocolState = TALKING
  },
  TALKING {
    override fun signal(): ProtocolState = WAITING
  };

  abstract fun signal(): ProtocolState
}

// Enum w/ companion static method is optimized
// CHECK-LABEL: class: redex.F
// CHECK-NEXT: Access flags:
// PRECHECK-NEXT: Superclass: java.lang.Enum
// POSTCHECK-NEXT: Superclass: java.lang.Object
enum class F {
  F0,
  F1,
  F2;
  companion object {
    fun getFByName(name: String): F = valueOf(name.toUpperCase())
  }
}

class Helper {
  companion object {
    // This method cannot be in the Test class to avoid Proguard keep rules.
    fun useEnumA(elem: A): Int =
        when (elem) {
          A.A0 -> 40
          A.A2 -> 42
          else -> 41
        }

    fun useEnumB(elem: B): Int =
        when (elem) {
          B.B1 -> 41
          B.B2 -> 42
          else -> 40
        }

    fun useEnumC(elem: C): Int =
        when (elem) {
          C.C0 -> 40
          C.C1 -> 41
          else -> 42
        }

    fun useEnumD(elem: D): String =
        when (elem) {
          D.D0 -> "Tool"
          D.D1 -> "APC"
          else -> "Puscifer"
        }

    fun useEnumE(elem: E): String =
        when (elem) {
          E.E0 -> "Hi"
          E.E1 -> "Yo"
          else -> "There"
        }

    fun useEnumProtocolState(elem: ProtocolState): String =
        when (elem.signal()) {
          ProtocolState.TALKING -> "waiting"
          else -> "talking"
        }

    fun useEnumF(name: String): String =
        when (F.Companion.getFByName(name)) {
          F.F0 -> "F0"
          F.F1 -> "F1"
          else -> "F2"
        }
  }
}

class EnumTransformTest {

  @Test
  fun testA() {
    assertThat(Helper.Companion.useEnumA(A.A0)).isEqualTo(40)
    assertThat(Helper.Companion.useEnumA(A.A1)).isEqualTo(41)
    assertThat(Helper.Companion.useEnumA(A.A2)).isEqualTo(42)
  }

  @Test
  fun testB() {
    assertThat(Helper.Companion.useEnumB(B.B0)).isEqualTo(40)
    assertThat(Helper.Companion.useEnumB(B.B1)).isEqualTo(41)
    assertThat(Helper.Companion.useEnumB(B.B2)).isEqualTo(42)
  }

  @Test
  fun testC() {
    assertThat(Helper.Companion.useEnumC(C.C0)).isEqualTo(40)
    assertThat(Helper.Companion.useEnumC(C.C1)).isEqualTo(41)
    assertThat(Helper.Companion.useEnumC(C.C2)).isEqualTo(42)
  }

  @Test
  fun testD() {
    assertThat(Helper.Companion.useEnumD(D.D0)).isEqualTo("Tool")
    assertThat(Helper.Companion.useEnumD(D.D1)).isEqualTo("APC")
    assertThat(Helper.Companion.useEnumD(D.D2)).isEqualTo("Puscifer")
  }

  @Test
  fun testE() {
    assertThat(Helper.Companion.useEnumE(E.E0)).isEqualTo("Hi")
    assertThat(Helper.Companion.useEnumE(E.E1)).isEqualTo("Yo")
    assertThat(Helper.Companion.useEnumE(E.E2)).isEqualTo("There")
  }

  @Test
  fun testProtocolState() {
    assertThat(Helper.Companion.useEnumProtocolState(ProtocolState.TALKING)).isEqualTo("talking")
  }

  @Test
  fun testF() {
    assertThat(Helper.Companion.useEnumF("f0")).isEqualTo("F0")
    assertThat(Helper.Companion.useEnumF("f1")).isEqualTo("F1")
    assertThat(Helper.Companion.useEnumF("f2")).isEqualTo("F2")
  }
}
